embedding/embedlite/embedshared/EmbedLiteViewBaseChild.cpp

```diff
@@ -1110,17 +1110,58 @@ mozilla::ipc::IPCResult EmbedLiteViewBaseChild::RecvHandleTextEvent(const nsStri
   return IPC_OK();
 }
 
+static KeyNameIndex getKeyNameIndexByDomKeyCode(int domKeyCode)
+{
+#define KEY(key_, _codeNameIdx, _keyCode, _modifier)
+#define CONTROL(keyNameIdx_, _codeNameIdx, _keyCode) \
+  if (domKeyCode == _keyCode) return KEY_NAME_INDEX_##keyNameIdx_;
+#include "KeyCodeConsensus_En_US.h"
+  return KEY_NAME_INDEX_USE_STRING;
+#undef CONTROL
+#undef KEY
+}
+
+static CodeNameIndex getCodeNameIndexByCharCode(int charCode)
+{
+  switch(charCode) {
+#define KEY(key_, _codeNameIdx, _keyCode, _modifier) \
+  case key_[0]: return CODE_NAME_INDEX_##_codeNameIdx;
+#define CONTROL(keyNameIdx_, _codeNameIdx, _keyCode)
+#include "KeyCodeConsensus_En_US.h"
+    default: return CODE_NAME_INDEX_UNKNOWN;
+#undef CONTROL
+#undef KEY
+  }
+}
+
+static Modifiers getModifiersByCharCode(int charCode)
+{
+  switch(charCode) {
+#define KEY(key_, _codeNameIdx, _keyCode, _modifier) \
+  case key_[0]: return _modifier;
+#define CONTROL(keyNameIdx_, _codeNameIdx, _keyCode)
+#include "KeyCodeConsensus_En_US.h"
+    default: return 0;
+#undef CONTROL
+#undef KEY
+  }
+}
+
 nsresult EmbedLiteViewBaseChild::DispatchKeyPressEvent(nsIWidget *widget, const EventMessage &message, const int &domKeyCode, const int &gmodifiers, const int &charCode)
 {
   NS_ENSURE_TRUE(widget, NS_ERROR_FAILURE);
   EventMessage msg = message;
   WidgetKeyboardEvent event(true, msg, widget);
-  event.mModifiers = Modifiers(gmodifiers);
+  event.mModifiers = Modifiers(gmodifiers) | getModifiersByCharCode(charCode);
   event.mKeyCode = charCode ? 0 : domKeyCode;
   event.mCharCode = charCode;
   event.mLocation = eKeyLocationStandard;
   event.mRefPoint = LayoutDeviceIntPoint(0, 0);
   event.mTime = PR_IntervalNow();
+  event.mKeyNameIndex = getKeyNameIndexByDomKeyCode(domKeyCode);
+  event.mKeyValue.Assign(charCode);
+  event.mCodeNameIndex = getCodeNameIndexByCharCode(charCode);
```

Comment:
Have you checked https://github.com/sailfishos/gecko-dev/blob/master/widget/NativeKeyToDOMCodeName.h and https://github.com/sailfishos/gecko-dev/blob/master/widget/NativeKeyToDOMKeyName.h ?

I'd check KeymapWrapper::InitKeyEvent from nsGtkKeyUtils.cpp

Reply:
CODE_MAP_X11 from NativeKeyToDOMCodeName.h cannot be used as it requires scancode, but only charCode is provided for character input (in QMozViewPrivate::inputMethodEvent). Also, not critical, but NativeKeyToDOMCodeName.h does not contain MODIFIER_SHIFT matches.
NativeKeyToDOMKeyName.h contains mapping between Qt::Key and widget::KeyNameIndex. Qt::Key in our case is available only on the qtmozembed side. So I also don't see a convenient way to use it.

Reply:
We could use Qt in embedding/embedlite if that help even though target is to get rid using it.